Updates Plugin Development Docs to use Poetry (#237)
nautobot/nautobot: nautobot/docs/plugins/development.md

* Establish their own REST API endpoints
* Add custom request/response middleware

Keep in mind that each piece of functionality is entirely optional. For example, if your plugin merely adds a piece of middleware or an API endpoint for existing data, there's no need to define any new models.

## Initial Setup

!!! important "Use a Development Environment, Not Production For Plugin Development"

>     The development environment should not be your production environment. For information on getting started with a development environment, check out [https://nautobot.readthedocs.io/en/latest/development/getting-started/](https://nautobot.readthedocs.io/en/latest/development/getting-started/).

Reviewer: ```suggestion
    You should not use your production environment for plugin development. For information on getting started with a development environment, check out [https://nautobot.readthedocs.io/en/latest/development/getting-started/](https://nautobot.readthedocs.io/en/latest/development/getting-started/).
```

### Create pyproject.toml

#### Poetry Init (Recommended)

To get started with a project using [Python Poetry](https://python-poetry.org/) you use the `poetry init` command. This will guide you through the prompts necessary to generate a pyproject.toml with details required for packaging.

```
This command will guide you through creating your pyproject.toml config.

Package name [tmp]:  nautobot-animal-sounds
Version [0.1.0]:
Description []:  An example Nautobot plugin
Author [, n to skip]:  Jeremy Stretch
License []:  Apache 2.0
Compatible Python versions [^3.8]:  ^3.6

Would you like to define your main dependencies interactively? (yes/no) [yes] no
Would you like to define your development dependencies interactively? (yes/no) [yes] no
Generated file

[tool.poetry]
name = "nautobot-animal-sounds"
version = "0.1.0"
description = "An example Nautobot plugin"
authors = ["Jeremy Stretch"]
license = "Apache 2.0"

[tool.poetry.dependencies]
python = "^3.6"

[tool.poetry.dev-dependencies]

[build-system]
requires = ["poetry-core>=1.0.0"]
build-backend = "poetry.core.masonry.api"

Do you confirm generation? (yes/no) [yes]
```

#### Manual

> The `pyproject.toml` is the recommended Python packaging methodology specified by [PEP 517](https://www.python.org/dev/peps/pep-0517/). If not able to build via Python Poetry (recommended method), the ability to package in the legacy `setup.py` methodology is still available, but not the recommended (and thus not documented) method to setup the package. An example pyproject.toml for setuptools is below, with the documentation on how to build with setuptools can be found on the [setuptools documentation](https://setuptools.readthedocs.io/en/latest/build_meta.html).

Reviewer: ```suggestion
The `pyproject.toml` is the recommended Python packaging methodology specified by [PEP 517](https://www.python.org/dev/peps/pep-0517/) and [PEP 518](https://www.python.org/dev/peps/pep-0518/). While the legacy `setup.py` Python packaging methodology is still available, it is not the recommended method to set up a package, and so it is not documented here. 

If for some reason you do not wish to use Poetry, `pyproject.toml` can be defined to use other packaging libraries. An example `pyproject.toml` for setuptools is below, and the documentation on how to build with setuptools can be found on the [setuptools documentation](https://setuptools.readthedocs.io/en/latest/build_meta.html).
```

Maybe move the first paragraph up to somewhere around line 59?

```toml
[build-system]
requires = ["setuptools", "wheel"]
build-backend = "setuptools.build_meta"

[metadata]
name = "nautobot-animal-sounds"
version = "0.1.0"
description = "An example Nautobot plugin"

> authors = ["Jeremy Stretch"]

Reviewer: ```suggestion
authors = ["Josh VanDeraa"]
```

Package name [tmp]:  nautobot-animal-sounds
Version [0.1.0]:
Description []:  An example Nautobot plugin

> Author [, n to skip]:  Josh VanDeraa

Reviewer 2: 😎

name = "nautobot-animal-sounds"
version = "0.1.0"
description = "An example Nautobot plugin"

> authors = ["Josh VanDeraa"]

Reviewer 2: 🕶️

| `jobs` | The dotted path to the list of Job classes (default: `jobs.jobs`) |
| `max_version` | Maximum version of Nautobot with which the plugin is compatible |
| `menu_items` | The dotted path to the list of menu items provided by the plugin (default: `navigation.menu_items`) |
| `middleware` | A list of middleware classes to append after Nautobot's built-in middleware |
| `min_version` | Minimum version of Nautobot with which the plugin is compatible |
| `name` | Raw plugin name; same as the plugin's source directory |
| `required_settings` | A list of any configuration parameters that **must** be defined by the user |
| `template_extensions` | The dotted path to the list of template extension classes (default: `template_content.template_extensions`) |
| `verbose_name` | Human-friendly name for the plugin |
| `version` | Current release ([semantic versioning](https://semver.org/) is encouraged) |

All required settings must be configured by the user. If a configuration parameter is listed in both `required_settings` and `default_settings`, the default setting will be ignored.

### Install the Plugin for Development

> To ease development, it is recommended with the Nautobot virtual environment activated (`/opt/nautobot/bin/activate` if using official doc). Then while in the location of the `pyproject.toml` file execute the command `poetry install` to install the local package. By default Poetry installs the current project as editable and the dependencies as well.

Reviewer 2: The first two sentences are confusing to me because we're saying to activate the virtualenv that appears to be in the production-deployed path of `/opt/nautobot`, but then also directing the reader to run `poetry install`?  Can you please clarify this?

The final sentence on a 2nd read through is a bit hard to grok. Also, some might not know what "editable" mode means. Can we reword this, please? Here's a suggestion:

```suggestion
To ease development, it is recommended with the Nautobot virtual environment activated (`/opt/nautobot/bin/activate` if using official doc). Then while in the location of the `pyproject.toml` file execute the command `poetry install` to install the local package. Poetry installs the current project and its dependencies in editable mode (aka ["development mode"](https://setuptools.readthedocs.io/en/latest/userguide/development_mode.html)).
```

The configurable attributes for a `PluginConfig` are listed below in alphabetical order.

> | Name                  | Description                                                                                                                                                                             |

Reviewer: I don't see a big need to reformat this table - did you make any actual content changes here?

Author: No changes. It formats nicer to read, except on small screens. Reverting.

| `required_settings`   | A list of any configuration parameters that **must** be defined by the user                                                                                                             |
| `template_extensions` | The dotted path to the list of template extension classes (default: `template_content.template_extensions`)                                                                             |
| `verbose_name`        | Human-friendly name for the plugin                                                                                                                                                      |
| `version`             | Current release ([semantic versioning](https://semver.org/) is encouraged)                                                                                                              |

All required settings must be configured by the user. If a configuration parameter is listed in both `required_settings` and `default_settings`, the default setting will be ignored.

### Install the Plugin for Development

> To ease development, it is recommended with the virtual environment activated to use `poetry install` command from the location of the `pyproject.toml` file. By default Poetry installs the current project as editable and the dependencies as well.

Reviewer: This could do with some clarification - what virtual environment is being referred to here?

```no-highlight
$ poetry install
```

> Once the Python package being developed is installed, add the plugin to the plugin configuration for Nautobot:

Reviewer: ```suggestion
Once the plugin has been installed, add it to the plugin configuration for Nautobot:
```

### Install the Plugin for Development

To install the plugin for development the following steps should be taken:

* Activate the Nautobot virtual environment (`/opt/nautobot/bin/activate` if using official doc)
* Navigate to the project root, where the `pyproject.toml` file exists for the plugin
* Execute the command `poetry install` to install the local package into the Nautobot virtual environment

!!! note
    Poetry installs the current project and its dependencies in editable mode (aka ["development mode"](https://setuptools.readthedocs.io/en/latest/userguide/development_mode.html)).

!!! important "This should be done in development environment"

>     The development environment should not be your production environment. For information on getting started with a development environment, check out [https://nautobot.readthedocs.io/en/latest/development/getting-started/](https://nautobot.readthedocs.io/en/latest/development/getting-started/).

Reviewer: ```suggestion
    You should not use your production environment for plugin development. For information on getting started with a development environment, check out [https://nautobot.readthedocs.io/en/latest/development/getting-started/](https://nautobot.readthedocs.io/en/latest/development/getting-started/).
```

### Install the Plugin for Development

The plugin needs to be installed into the same python environment where Nautobot is, so that we can get access to `nautobot-server` command, and also so that the nautobot-server is aware of the new plugin.  

> If you installed Nautobot using poetry, then go to the Nautobot repo clone and run `poetry shell` there.  

Reviewer 2: ```suggestion
If you installed Nautobot using Poetry, then go to the root directory of your clone of the Nautobot repository and run `poetry shell` there.  Afterward, return to the root directory of your plugin to continue development.
```

> Otherwise, manually activate nautobot using `source /opt/nautobot/bin/activate`.

Reviewer 2: ```suggestion
Otherwise if you're using the Docker workflow, manually activate the Nautobot environment using `source /opt/nautobot/bin/activate`.
```

To install the plugin for development the following steps should be taken:

> * Activate the Nautobot virtual environment (`/opt/nautobot/bin/activate` if using official doc)

Reviewer 2: ```suggestion
* Activate the Nautobot virtual environment (as detailed above)
```

        return self.name
```

Once you have defined the model(s) for your plugin, you'll need to create the database schema migrations. A migration file is essentially a set of instructions for manipulating the database to support your new model, or to alter existing models.  

Creating migrations can be done automatically using the `nautobot-server makemigrations <plugin_name>` management command, where `<plugin_name>` is the name of the Python package for your plugin (e.g. `animal_sounds`):

```no-highlight

> $ nautobot-server makemigrations animal_sounds

Reviewer: Do we have some inconsistencies on the plugin name/slug in this document? In the below example output it's `nautobot_animal_sounds` but here and above you're using `animal_sounds`.

        return self.name
```

> Once you have defined the model(s) for your plugin, you'll need to create the database schema migrations. A migration file is essentially a set of instructions for manipulating the PostgreSQL database to support your new model, or to alter existing models. Creating migrations can usually be done automatically using Django's `nautobot-server makemigrations <plugin_name>` management command.

Reviewer: ```suggestion
Once you have defined the model(s) for your plugin, you'll need to create the database schema migrations. A migration file is essentially a set of instructions for manipulating the PostgreSQL database to support your new model, or to alter existing models. Creating migrations can usually be done automatically using the `nautobot-server makemigrations <plugin_name>` command.
```

        return self.name
```

> Once you have defined the model(s) for your plugin, you'll need to create the database schema migrations. A migration file is essentially a set of instructions for manipulating the PostgreSQL database to support your new model, or to alter existing models. Creating migrations can usually be done automatically using the `nautobot-server makemigrations <plugin_name>` command.

Reviewer 2: ```suggestion
Once you have defined the model(s) for your plugin, you'll need to create the database schema migrations. A migration file is essentially a set of instructions for manipulating the database to support your new model, or to alter existing models. 

Creating migrations ca be done automatically using the `nautobot-server makemigrations <plugin_name>` management command, where `<plugin_name>` is the name of the Python package for your plugin (e.g. `animal_sounds`):

```no-highlight
$ nautobot-server makemigrations animal_sounds
# ``` <= Suggestion is screwed up because of backticks.
```

Reviewer 2: Don't commit this suggestion, since the formatting got hosed! :)

![Nautobot plugin in the admin UI](../media/plugins/plugin_admin_ui.png)

### Register the Model in the GraphQL Interface

> Plugins can optionally expose their models via the GraphQL interface to allow the models to be part of the Graph and to be queried easily. There are two ways to expose a model to the graphql interface.
> * By using the `extras_features` decorator
> * By creating your own GraphQL Type object and registering it within `graphql/types.py` of your plugin (decorator is not needed)

Reviewer 3: ```suggestion
Plugins can optionally expose their models via the GraphQL interface to allow the models to be part of the Graph and to be queried easily. There are two ways to expose a model to the graphql interface.

* By using the `extras_features` decorator
* By creating your own GraphQL Type object and registering it within `graphql/types.py` of your plugin (decorator is not needed)
```

![Nautobot plugin in the admin UI](../media/plugins/plugin_admin_ui.png)

### Register the Model in the GraphQL Interface

Plugins can optionally expose their models via the GraphQL interface to allow the models to be part of the Graph and to be queried easily. There are two ways to expose a model to the graphql interface.

* By using the `extras_features` decorator
* By creating your own GraphQL Type object and registering it within `graphql/types.py` of your plugin (decorator is not needed)

> #### Using the `extras_features` Decorator for Graphql

Reviewer 2: ```suggestion
#### Using the `extras_features` Decorator for GraphQL
```
